pkg/nodeconfig/nodeconfig.go

```diff
@@ -496,6 +612,97 @@ func (nc *nodeConfig) configureWICD() error {
 	return nc.Windows.ConfigureWICD(nodeConfigCache.apiServerEndpoint, saCA, saToken)
 }
 
+// generateKubeconfig creates a kubeconfig spec with the certificate and token data from the given secret
+func generateKubeconfig(secret *core.Secret, apiServerURL string) (clientcmdv1.Config, error) {
+	// extract ca.crt and token data fields
+	caCert := secret.Data[core.ServiceAccountRootCAKey]
+	if caCert == nil {
+		return clientcmdv1.Config{}, errors.Errorf("unable to find %s CA cert in secret %s",
+			core.ServiceAccountRootCAKey, secret.GetName())
+	}
+	token := secret.Data[core.ServiceAccountTokenKey]
+	if token == nil {
+		return clientcmdv1.Config{}, errors.Errorf("unable to find %s token in secret %s",
+			core.ServiceAccountTokenKey, secret.GetName())
+	}
+	kubeconfig := clientcmdv1.Config{
+		Clusters: []clientcmdv1.NamedCluster{{
+			Name: "local",
+			Cluster: clientcmdv1.Cluster{
+				Server:                   apiServerURL,
+				CertificateAuthorityData: caCert,
+			}},
+		},
+		AuthInfos: []clientcmdv1.NamedAuthInfo{{
+			Name: "kubelet",
+			AuthInfo: clientcmdv1.AuthInfo{
+				Token: string(token),
+			},
+		}},
+		Contexts: []clientcmdv1.NamedContext{{
+			Name: "kubelet",
+			Context: clientcmdv1.Context{
+				Cluster:  "local",
+				AuthInfo: "kubelet",
+			},
+		}},
+		CurrentContext: "kubelet",
+	}
+	return kubeconfig, nil
+}
+
+// generateKubeletConfiguration returns the configuration spec for the kubelet Windows service
+func generateKubeletConfiguration(clusterDNS string) kubeletconfig.KubeletConfiguration {
```

Comment:
Consider adding a unit test for this function.

Reply:
I'm not convinced on the unit testing of this or the kubeconfig. The test for this would be reconstructing the `KubeletConfiguration` object with the same values as in this function, then asserting they are equal. Don't see much benefit

Reply:
Unit test added after further discussion